stan/math/opencl/triangular.hpp

```diff
@@ -0,0 +1,133 @@
+#ifndef STAN_MATH_OPENCL_TRIANGULAR_HPP
+#define STAN_MATH_OPENCL_TRIANGULAR_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/stringify.hpp>
+#include <Eigen/Core>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+enum class TriangularViewCL { Diagonal = 0, Lower = 1, Upper = 2, Entire = 3 };
+
+/**
+ * Combines two triangular views. Result is nonzero, where any of the inputs is
+ * nonzero.
+ * @param a first view
+ * @param b second view
+ * @return combined view
+ */
+inline TriangularViewCL combine(TriangularViewCL a, TriangularViewCL b) {
+  typedef typename std::underlying_type<TriangularViewCL>::type underlying;
+  return static_cast<TriangularViewCL>(static_cast<underlying>(a)
+                                       | static_cast<underlying>(b));
```

Comment:
Is your intent here to look at the two types? I'm not seeing why we don't just do some `if`s? These are defined as `int`s so I'm not really getting what the `|` is gonna do here

Reply:
I will try to explain this the long way. Hopefully it will include whatever is confusing here. In following explanation im prefixing binary numbers with b.

TriangularViewCL is a bitfield. Only the lowest two bits the underlying integral type are used - all other are always zero. The lowest bit carries information on whether upper triangular part of a matrix is zero (value 1) or nonzero (value 0). So Diagonal(0 = b00) and Upper (2 = b10) have this bit set to 0. Lower(1 = b01) and Entire(3 = b11) have this bit set to 1.

In the same way the second lowest bit carries the information about lower triangular part. Again 0 stands for any nonzero values and 1 stands for all zero. So Diagonal(0 = b00) and Lower(1 = b01) have this set to 0. Upper (2 = b10) and Entire(3 = b11) have this bit set to 1.

`|` is bitwise or operation. Each bit of the result is 1 that bit of at least one inputs is 1. `&` is bitwise and operation. Each bit of the result is 1 if that bit of both inputs is 1. `~` is bitwise negation. Each bit of the result is 1 if that bit of the input is 0.

Some examples: 
Entire | Lower = b11 | b01 = b11 = Entire
Upper | Lower = b10 | b01 = b11 = Entire
Lower | Diagonal = b01 | b00 = b01 = Lower
~Lower = ~ b01 = b10 = Upper
~Diagonal = ~ b00 = b11 = Entire
Lower & Upper = b01 & b10 = b00 = Diagonal

The last one even reads as english: If a matrix is lower triangular and upper triangular it is diagonal.

Since I have seen it written a few times that we should assume that whoever is reading the code is proficient in C++ I prefere it this way compared to many ifs.

Reply:
Thanks! After playing around with the code I get that now. I didn't know that C++ would treat `int`s as bitfields and assumed it would try to deduce them to boolean which was my original confusion.

With the above it makes sense, if you know the order of the `enum`. But without that info then `Upper & Lower` is still hard to read. The best way to put it is to imagine you are coming into the codebase with no knowledge and see `Upper | Lower`. You would have to go find the `enum` in order to deduce the output. 

I think if we change the `+` and `*` operators to `union` and `intersection` that will both convey the actual implementation and be more readable without knowing the underlying implementation. Does that work for you?

Reply:
Yes.